platform/kube/register.go

```diff
@@ -0,0 +1,196 @@
+	// For most common ports allow the protocol to be guessed, this isn't meant
+	// to replace /etc/services. Fully qualified proto[-extra]:port is the
+	// recommended usage.
+	portsToName = map[int32]string{
+		80:   "http",
+		443:  "https",
+		3306: "mysql",
+		8080: "http",
+	}
+)
+
+// NamedPort defines the Port and Name tuple needed for services and endpoints.
+type NamedPort struct {
+	Port int32
+	Name string
+}
+
+// Str2NamedPort parses a proto:port string into a namePort struct.
+func Str2NamedPort(str string) (NamedPort, error) {
+	var r NamedPort
+	idx := strings.Index(str, ":")
+	if idx >= 0 {
+		r.Name = str[:idx]
+		str = str[idx+1:]
+	}
+	p, err := strconv.Atoi(str)
+	if err != nil {
+		return r, err
+	}
+	r.Port = int32(p)
+	if len(r.Name) == 0 {
+		name, found := portsToName[r.Port]
+		r.Name = name
+		if !found {
+			r.Name = str
+		}
+	}
+	return r, nil
+}
+
+// samePorts returns true if the numerical part of the ports is the same.
+// The arrays aren't necessarily sorted so we (re)use a map.
+func samePorts(ep []v1.EndpointPort, portsMap map[int32]bool) bool {
+	if len(ep) != len(portsMap) {
+		return false
+	}
+	for _, e := range ep {
+		if !portsMap[e.Port] {
+			return false
+		}
+	}
+	return true
+}
+
+// splitEqual splits key=value string into key,value. if no = is found
+// the whole string is the key and value is empty.
+func splitEqual(str string) (string, string) {
+	idx := strings.Index(str, "=")
+	var k string
+	var v string
+	if idx >= 0 {
+		k = str[:idx]
+		v = str[idx+1:]
+	} else {
+		k = str
+	}
+	return k, v
+}
+
+// addLabelsAndAnnotations adds labels and annotations to an object.
+func addLabelsAndAnnotations(obj *meta_v1.ObjectMeta, labels []string, annotations []string) {
+	if obj.Labels == nil {
+		obj.Labels = make(map[string]string, len(labels))
+	}
+	for _, l := range labels {
+		k, v := splitEqual(l)
+		obj.Labels[k] = v
+	}
+	if obj.Annotations == nil {
+		obj.Annotations = make(map[string]string, len(annotations))
+	}
+	for _, a := range annotations {
+		k, v := splitEqual(a)
+		obj.Annotations[k] = v
+	}
+}
+
+// RegisterEndpoint registers the endpoint (and the service if it doesn't
+// already exists). It creates or updates as needed. When creating it adds the
+// optional labels.
+func RegisterEndpoint(client kubernetes.Interface, namespace string, svcName string,
+	ip string, portsList []NamedPort, labels []string, annotations []string) error {
+	getOpt := meta_v1.GetOptions{IncludeUninitialized: true}
+	_, err := client.Core().Services(namespace).Get(svcName, getOpt)
+	if err != nil {
+		glog.Warningf("Got '%v' looking up svc '%s' in namespace '%s', attempting to create it", err, svcName, namespace)
+		svc := v1.Service{}
+		svc.Name = svcName
+		for _, p := range portsList {
+			svc.Spec.Ports = append(svc.Spec.Ports, v1.ServicePort{Name: p.Name, Port: p.Port})
+		}
+		addLabelsAndAnnotations(&svc.ObjectMeta, labels, annotations)
+		_, err = client.CoreV1().Services(namespace).Create(&svc)
```

Comment:
To your comment:
```I think the UX matches the 0.2 use case. this is also why creating the svc is a warning btw - I could add a flag to make it an optional error but I think that just gets in the way of the user for no reason ? (which pr?)```

Simplifying endpoint registration is fine.. But allowing a CLI tool to create service, along with the endpoint is very risky in production. If we want people to confidently use this anywhere in production, we should restrict ourselves to the minimal write permissions required in kube API server from a VM -- creating the Endpoint.

A compromised VM would allow the attacker to define an infinite number of services or possibly register the VM with a k8s service that already exists?

Reply:
throwing another reason: if there is an even small scripting error such that I programmatically populate svc to be 
```
for i in seq 1 100
do
istioctl register $i ${VMs[$i]} 8080
done
``` 
imagine the consequences..

Reply:
this tool isn't (necessarily) meant to be run from the VM itself, it's probably best to be run by the k8s admin

I'm not going to debate that we shouldn't have convenient tools because people can make mistakes... you can put kubectl in a loop too